kotlin-eclipse-ui/src/org/jetbrains/kotlin/ui/editors/hover/KotlinJavadocTextHover.kt

```diff
@@ -16,52 +16,198 @@
+
     override fun getHoverInfo(hoverData: HoverData): BrowserInformationControlInput? {
-        val (element, editor) = hoverData
-        
-        val javaElements = obtainJavaElements(element)
-        if (javaElements.isEmpty()) return null
-        
-        val editorInputElement = EditorUtility.getEditorInputJavaElement(editor.javaEditor, false)
-        
-        return JavadocHover.getHoverInfo(javaElements.toTypedArray(), editorInputElement, DUMMY_REGION, null)
+        val (element, _) = hoverData
+        return KotlinAnnotationBrowserInformationControlInput(element)
     }
-    
+
     override fun isAvailable(hoverData: HoverData): Boolean = true
-    
-    override fun getHoverControlCreator(editor: KotlinEditor): IInformationControlCreator? {
-        return JavadocHover.PresenterControlCreator(editor.javaEditor.getSite())
-    }
-    
-    private fun obtainJavaElements(element: KtElement): List<IJavaElement> {
-        val expression = getReferenceExpression(element) ?: return emptyList()
-        
-        return createReferences(expression)
-                .resolveToSourceElements()
-                .filterIsInstance(EclipseJavaSourceElement::class.java)
-                .flatMap { it.toJavaElements() }
+
+    override fun getHoverControlCreator(editor: KotlinEditor): IInformationControlCreator? =
+        JavadocHover.PresenterControlCreator(editor.javaEditor.site)
+
+    private class KotlinAnnotationBrowserInformationControlInput(
+        val element: KtElement,
+        previousInput: BrowserInformationControlInput? = null
+    ) : BrowserInformationControlInput(previousInput) {
+
+        companion object {
+            const val imageWidth = 16
+            const val imageHeight = 16
+            const val labelLeft = 20
+            const val labelTop = 2
+            val javadocStyleSheet: String by lazy { loadStyleSheet("/JavadocHoverStyleSheet.css") }
+            val registry: ColorRegistry by lazy { JFaceResources.getColorRegistry() }
+            val fgRGB: RGB by lazy { registry.getRGB("org.eclipse.jdt.ui.Javadoc.foregroundColor") }
+            val bgRGB: RGB by lazy { registry.getRGB("org.eclipse.jdt.ui.Javadoc.backgroundColor") }
+
+            fun wrappedHeader(content: String) =
+                "<h5><div style='word-wrap: break-word; position: relative; margin-left: ${labelLeft}px; padding-top: ${labelTop}px;'>$content</div></h5>"
+
+            fun img(src: URL) =
+                "<img style='border:none; position: absolute; width: ${imageWidth}px; height: ${imageHeight}px; left: -${labelLeft + 1}px;' src='$src'/>"
+        }
+
+        private val expression = getReferenceExpression(element)
+
+        private val descriptor = expression?.let {
+            KotlinAnalysisFileCache.getAnalysisResult(element.containingKtFile).analysisResult.bindingContext[REFERENCE_TARGET, expression]
+        } ?: (element as? KtDeclaration)?.resolveToDescriptor()
```

Comment:
Seems quite complicated to me, why not just 
```
(expression ?: element as? KtDeclaration)?.resolveToDescriptor()
```
?

Reply:
Actually, resolveToDescriptor browses DECLARATION_TO_DESCRIPTOR Slice and I need to check REFERENCE_TARGET one instead.